<launch> <!-- Version 2.0 -->

<arg name="database_folder" default="$(find iamech_ros)/maps/"/>
<arg name="database_name" default="rs_lidar_map"/>
<arg name="namespace" default='iAmech'/>

<!-- Realsense Data Topic-->
<arg name="depth_topic" default="/camera/aligned_depth_to_color/image_raw"/>
<arg name="rgb_topic" default="/camera/color/image_raw"/>
<arg name="camera_info" default="/camera/color/camera_info"/>

<!-- 2D LiDAR Data Topic -->
<arg name="scan_topic" default="/scan"/>

<!-- Mapping Node -->
<group ns="rtabmap">
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="--delete_db_on_start --Optimizer/GravitySigma 0.3">

        <!-- Basic RTAB-Map Parameters -->
        <param name="database_path"       type="string" value="$(arg database_folder)$(arg database_name).db"/>
        <param name="frame_id"            type="string" value="base_link"/>
        <param name="odom_frame_id"       type="string" value="odom"/>
        <param name="subscribe_depth"     type="bool"   value="true"/>
        <param name="subscribe_rgb"     type="bool"   value="true"/>
        <param name="wait_imu_to_init"    type="bool"   value="false"/>
        <param name="subscribe_scan"      type="bool"   value="true"/>

        <!-- RTAB-Map Inputs -->
        <remap from="scan" to="$(arg scan_topic)"/>
        <remap from="rgb/image" to="$(arg rgb_topic)"/>
        <remap from="depth/image" to="$(arg depth_topic)"/>
        <remap from="rgb/camera_info" to="$(arg camera_info)"/>

        <!-- RTAB-Map Output -->
        <remap from="grid_map" to="/map"/>

        <param name="queue_size" type="int" value="10"/>

        <!-- RTAB-Map's parameters -->
        <param name="Rtabmap/DetectionRate" type="string" value="1"/> <!-- Rate (Hz) at which new nodes are added to map -->
        <param name="RGBD/NeighborLinkRefining" type="string" value="true"/>
        <param name="RGBD/ProximityBySpace"     type="string" value="true"/>
        <param name="RGBD/AngularUpdate"        type="string" value="0.01"/>
        <param name="RGBD/LinearUpdate"         type="string" value="0.01"/>
        <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/>
        <param name="Grid/Sensor"               type="string" value="0"/> <!-- false: occupancy grid from lidar -->
        <param name="Reg/Force3DoF"             type="string" value="true"/> <!-- True : 2D SLAM, False: 3D SLAM -->
        <param name="Mem/NotLinkedNodesKept" type="string" value="false"/> <!-- Set to false to avoid saving data when robot is not moving -->

        <!-- Loop Closure Constraint --> 
        <param name="Reg/Strategy"              type="string" value="0"/> <!-- 0=Visual, 1=ICP (1 requires scan), 2=Visual+ICP-->
        <param name="Vis/MinInliers" type="string" value="15"/> <!-- Minimum visual inliers to accept loop closure -->
        <param name="RGBD/ProximityPathMaxNeighbors" type="string" value="0"/>
             
        <!-- Loop Closure Detection -->
        <param name="Kp/DetectorStrategy" type="string" value="2"/> <!-- 0=SURF 1=SIFT 2=ORB 3=FAST/FREAK 4=FAST/BRIEF 5=GFTT/FREAK 6=GFTT/BRIEF 7=BRISK 8=GFTT/ORB 9=KAZE -->
        <param name="Vis/FeatureType"                type="string" value="2"/>
        <param name="Kp/MaxFeatures" type="string" value="400"/> <!-- Maximum visual words per image (bag-of-words) -->
        <param name="SURF/HessianThreshold" type="string" value="100"/> <!-- Used to extract more or less SURF features -->

        <!-- ICP parameters -->
        <param name="Icp/VoxelSize"                 type="string" value="0.05"/>
        <param name="Icp/MaxCorrespondenceDistance" type="string" value="0.1"/>
        
        <param name="Mem/IncrementalMemory" type="string" value="true"/>
        <param name="Mem/InitWMWithAllNodes" type="string" value="false"/>

        <!-- Jessica's work -->
        <param name="Grid/MaxGroundHeight" type="string" value="-0.7"/>       <!-- all points below -70 cm are ground -->
        <param name="Grid/MaxObstacleHeight" type="string" value="0.5"/>       <!-- all points above -70 cm and below 50 m are obstacles -->
        <param name="Grid/NormalsSegmentation" type="string" value="false"/> <!-- Use simple passthrough to filter the ground instead of normals segmentation -->
    </node>
</group>

<node type="rviz" name="rviz" pkg="rviz" args="-d $(find iamech_ros)/rviz/mapping.rviz" />

</launch>
